src/set_theory/game/nim.lean

```diff
@@ -0,0 +1,283 @@
+  { intro j,
+    let hwf : (ordinal.typein O.out.r j) < O := nim_wf_lemma j,
+    rw move_right_mk,
+    exact nim_impartial (ordinal.typein O.out.r j) }
+end
+using_well_founded {dec_tac := tactic.assumption}
+
+lemma nim_zero_first_loses : (nim (0:ordinal)).first_loses :=
+begin
+  rw nim_def,
+  split;
+  rw le_def_lt;
+  split;
+  intro i;
+  try {rw left_moves_mk at i};
+  try {rw right_moves_mk at i};
+  try { have h := ordinal.typein_lt_type (quotient.out (0:ordinal)).r i,
+        rw ordinal.type_out at h,
+        have hcontra := ordinal.zero_le (ordinal.typein (quotient.out (0:ordinal)).r i),
+        have h := not_le_of_lt h,
+        contradiction };
+  try { exact pempty.elim i }
+end
+
+lemma nim_non_zero_first_wins (O : ordinal) (hO : O ≠ 0) : (nim O).first_wins :=
+begin
+  rw nim_def,
+  rw ←ordinal.pos_iff_ne_zero at hO,
+  split;
+  rw lt_def_le,
+  { left,
+    use (ordinal.principal_seg_out hO).top,
+    rw [move_left_mk, ordinal.typein_top, ordinal.type_out],
+    exact nim_zero_first_loses.2 },
+  { right,
+    use (ordinal.principal_seg_out hO).top,
+    rw [move_right_mk, ordinal.typein_top, ordinal.type_out],
+    exact nim_zero_first_loses.1 }
+end
+
+lemma nim_sum_first_loses_iff_eq (O₁ O₂ : ordinal) : (nim O₁ + nim O₂).first_loses ↔ O₁ = O₂ :=
+begin
+  split,
+  { contrapose,
+    intros hneq hp,
+    wlog h : O₁ ≤ O₂ using [O₁ O₂, O₂ O₁],
+    exact ordinal.le_total O₁ O₂,
+    { have h : O₁ < O₂ := lt_of_le_of_ne h hneq,
+      rw ←(no_good_left_moves_iff_first_loses $ impartial_add (nim_impartial O₁) (nim_impartial O₂))
+				at hp,
+      equiv_rw left_moves_add (nim O₁) (nim O₂) at hp,
+      rw nim_def O₂ at hp,
+      specialize hp (sum.inr (ordinal.principal_seg_out h).top),
+      rw [id, add_move_left_inr, move_left_mk, ordinal.typein_top, ordinal.type_out] at hp,
+      cases hp,
+      have hcontra := (impartial_add_self $ nim_impartial O₁).1,
+      rw ←pgame.not_lt at hcontra,
+      contradiction },
+    exact this (λ h, hneq h.symm) (first_loses_of_equiv add_comm_equiv hp) },
+  { intro h,
+    rw h,
+    exact impartial_add_self (nim_impartial O₂) }
+end
+
+lemma nim_sum_first_wins_iff_neq (O₁ O₂ : ordinal) : (nim O₁ + nim O₂).first_wins ↔ O₁ ≠ O₂ :=
+begin
+  split,
+  { intros hn heq,
+    cases hn,
+    rw ←nim_sum_first_loses_iff_eq at heq,
+    cases heq with h,
+    rw ←pgame.not_lt at h,
+    contradiction },
+  { contrapose,
+    intro hnp,
+    rw [not_not, ←nim_sum_first_loses_iff_eq],
+    cases impartial_winner_cases (impartial_add (nim_impartial O₁) (nim_impartial O₂)),
+    assumption,
+    contradiction }
+end
+
+/-- This definition will be used in the proof of the Sprague-Grundy theorem. It takes a function
+  from some type to ordinals and returns a nonempty set of ordinals with empty intersection with
+  the image of the function. It is guaranteed that the smallest ordinal not in the image will be
+  in the set, i.e. we can use this to find the mex -/
```

Comment:
mex?

Reply:
The mex is this: https://en.wikipedia.org/wiki/Mex_(mathematics)

The docstring seems to be ending without punctuation though.